File: 2023/Leetcode/problem69.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function(x) {
    if (x < 2) {
        return x;
    }

    let low = 0;
    let high = x;
    let median = Math.floor(x / 2);

    while (low != median) {
        if (median * median <= x) {
            low = median;
        } else {
            high = median;
        }
        median = Math.floor((low + high) / 2);
    }

    return median;
};

File: 2023/Leetcode/problem69.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x <= 1:
            return x

        low = 0
        high = x
        guess = (low + high) // 2

        while guess * guess != x and high - low > 1:
            if guess * guess < x:
                low = guess
            elif guess * guess > x:
                high = guess

            guess = (low + high) // 2

        if high * high == x:
            return high
        return guess